File: Python3/1 - 200/106 Construct Binary Tree from Inorder and Postorder Traversal.py
# 92ms, 82.79%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if not inorder: return None
        root = TreeNode(postorder[-1])
        stack = [root]
        i, j = len(postorder) - 2, len(inorder) - 1
        while i >= 0:
            node = TreeNode(postorder[i])
            i, tmp = i - 1, None
            while stack and stack[-1].val == inorder[j]:
                tmp, j = stack.pop(), j - 1
            if tmp: tmp.left = node
            else: stack[-1].right = node
            stack.append(node)
        return root
    

File: 1 - 200/106 Construct Binary Tree from Inorder and Postorder Traversal.cc
// 8ms, 94.57%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    TreeNode *buildTree(const vector<int> &inorder, const vector<int> &postorder) {
        if (inorder.empty()) return nullptr;

        TreeNode *root = new TreeNode(postorder.back());
        int i = postorder.size() - 2, j = inorder.size() - 1;
        stack<TreeNode *> st;
        st.push(root);

        while (i >= 0) {
            TreeNode *node = new TreeNode(postorder[i--]);
            TreeNode *tmp = nullptr;
            while (!st.empty() && st.top()->val == inorder[j]) {
                tmp = st.top();
                st.pop();
                --j;
            }
            (tmp ? tmp->left : st.top()->right) = node;
            st.push(node);
        }

        return root;
    }
};
